### PR TITLE
iosevka: 3.2.2 -> 3.7.1

### PR DESCRIPTION
**Pull request by AluisioASG**
###### Motivation for this change

Both packages were outdated and out of sync with one another.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/iosevka/bin.nix
Comment (midchildan) [inline pkgs/data/fonts/iosevka/bin.nix:4]: Do `iosevka` and `iosevka-bin` always have to be in sync with each other? If so, I think we should ensure they'd always stay the same by changing this line to `version = iosevka.version;`.

File: pkgs/data/fonts/iosevka/package.json
Comment (midchildan) [inline pkgs/data/fonts/iosevka/package.json:1]: Could you add a script that generates this automatically? Here's a working example:

```bash
#!/usr/bin/env bash

# Script to generate package.json needed to create the build environment.

set -eu -o pipefail
cd "$(dirname "${BASH_SOURCE[0]}")"

SRC="$(nix-build -A iosevka.src --no-out-link ../../../..)"
JQ_BIN="$(nix-build -A jq --no-out-link ../../../..)/bin/jq"


main() {
    jq '{
            "name": "iosevka-build-deps",
            "version": .version,
            "scripts": { "build": .scripts.build },
            "dependencies": .dependencies
        }' \
        "$SRC/package.json" \
        > package.json
}

jq() {
    command "$JQ_BIN" "$@"
}

main
```

**Review by midchildan — CHANGES_REQUESTED** [review]

File: pkgs/data/fonts/iosevka/bin.nix
Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/bin.nix:4]: They don't need to be, it just happens that `iosevka-bin` gets updated automatically, while `iosevka` needs the `package.json` to be regenerated. It'd make it less confusing if they were updated in lockstep though.

File: pkgs/data/fonts/iosevka/package.json
Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/package.json:1]: The latest push includes an `update.sh` script that does it.

**Comment by AluisioASG** [comment]
Made some changes:
- centralized version number and hashes in srcs.nix
- filtered package.json to keep only name, version, and dependencies
- added an update script

**Review by midchildan — CHANGES_REQUESTED** [review]
Great work! There's a single typo that needs fixing, but otherwise looks good.

File: pkgs/data/fonts/iosevka/update.sh
Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/update.sh]: Thanks!

**Review by ttuegel — APPROVED** [review]

**Review by midchildan — APPROVED** [review]

**Review by 414owen — APPROVED** [review]

**Review by rileyinman — APPROVED** [review]
Sorry it took me a bit to get around to this—the changes look great. The auto-update script especially will be handy, since the update process was a bit tedious and that's one of the reasons I hadn't gotten around to the new version yet. Thanks for the improvements!

**Review by midchildan — CHANGES_REQUESTED** [review]

**Review by midchildan — APPROVED** [review]

**Comment by AluisioASG** [comment]
Updated to 3.7.0.

@ryantm what's the current procedure for updating node-packages.nix? I've removed the update from this PR but I'm guessing it'll break the ofborg CI run (as well as the automated update once I stop calling generate.sh from the update script).

File: pkgs/data/fonts/iosevka/update.sh
Comment (doronbehar) [inline pkgs/data/fonts/iosevka/update.sh:39]: I think what I get from here is that it's not available on npm? That's why you can't put it in `node-packages.json`? If so, you should know it's possible to do this:

https://github.com/NixOS/nixpkgs/blob/a38590d91ce79e56d1bae79c5824008e906813ef/pkgs/development/node-packages/node-packages.json#L145

And you then will need only to update the version tag there every once in a while.. If that will be possible, you should be able to remove a lot of moving parts from these derivations.

**Comment by AluisioASG** [comment]
The latest commits are a significant rework based on @doronbehar's suggestion, at the expense of keeping the packages unsynchronized as they are today.

iosevka-bin can be updated with `genericUpdater`, but it needs a `src` attribute pointing to itself for that to work, because of

https://github.com/NixOS/nixpkgs/blob/242441fea2a8981dccf5800288c7ca02a1c31cac/pkgs/common-updater/scripts/update-source-version#L82-L83

iosevka's update script now only needs to change a URL in node-packages.json, but that update isn't propagated to the `iosevka` package until node-packages.nix is regenerated.

Since the two packages are again separated, this may be better off as two separate PRs. I'm marking this up as a draft so it doesn't get merged until at least I've cleaned up the commits.

File: pkgs/data/fonts/iosevka/update.sh
Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/update.sh]: Well, that does indeed simplify things a lot. I had to use a tarball though, as node2nix doesn't perform a shallow fetch and Iosevka's repo is 1GB+ in size.

File: pkgs/data/fonts/iosevka/bin.nix
Comment (doronbehar) [inline pkgs/data/fonts/iosevka/bin.nix:3]: What is the purpose of this package, and what does it provide that `iosevka` doesn't? Overall, I'm asking because I'd like to push this towards a minimal PR that only puts `iosevka` into `node-packages.json` to allow @cideM grab it to his next round of `./generate.sh`.

Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/bin.nix:3]: According to [the PR that inits it](https://github.com/NixOS/nixpkgs/pull/32106) and [one of the comments it points to](https://github.com/NixOS/nixpkgs/pull/31835#issuecomment-347048242), one of Iosevka's dependencies (otfcc) didn't build on aarch64 back then. However, I no longer see otfcc in the dependencies list (possibly replaced with [ot-builder](https://github.com/ot-builder/monorepo)), so it'd be worth checking if it works now, [which it seems like it does?](https://github.com/NixOS/nixpkgs/pull/99004/checks?check_run_id=1333017601)

Comment (AluisioASG) [reply pkgs/data/fonts/iosevka/bin.nix:3]: Oh, there is also [this](https://github.com/NixOS/nixpkgs/pull/101132) now.

**Comment by AluisioASG** [comment]
@doronbehar since #101132 took care of `iosevka-bin`, I removed the changes that dealt with it (but I moved its update script to make clear what package it is updating).

Should I drop the last commit (that updates node-packages.nix) even if it breaks `iosevka` on master in the meantime?

**Comment by doronbehar** [comment]
OK. It took me a while to grab my head around the changes. It looks good, although it's a bit of a mess. The problem is that many users are probably happy with iosevka-bin since it's easier to compile an override of it if one wishes only to get different / certain variants. Hence it's not suitable to discard the prebuild derivation for the sake of the source build derivation.

Unrelated, for iosevka, it's good to put it's deps in node-packages.json (as it used to be, only previously it was a manually downloaded package.json) so it will (hopefully) get updated frequently by the node packages maintainers (which we currently lack officially).

I was considering maybe suggesting to put the update script of the source build derivation in node-packages' `./generate.sh` but that seems like too much to me, and inappropriate though I'm open to argue about it.

I have a few nitpicks left though (appropriate to put in separate commits in this PR):

```diff
diff --git i/pkgs/data/fonts/iosevka/bin.nix w/pkgs/data/fonts/iosevka/bin.nix
index c3b7767bccc..853c338e184 100644
--- i/pkgs/data/fonts/iosevka/bin.nix
+++ w/pkgs/data/fonts/iosevka/bin.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchurl, unzip
+{ stdenv, lib, fetchurl, unzip, iosevka # Only for it's meta
 , variant ? ""
 }:
 
@@ -29,15 +29,7 @@ in stdenv.mkDerivation rec {
     unzip -d $out/share/fonts/truetype $src
   '';
 
-  meta = with lib; {
-    homepage = "https://be5invis.github.io/Iosevka/";
-    downloadPage = "https://github.com/be5invis/Iosevka/releases";
-    description = ''
-      Slender monospace sans-serif and slab-serif typeface inspired by Pragmata
-      Pro, M+ and PF DIN Mono, designed to be the ideal font for programming.
-    '';
-    license = licenses.ofl;
-    platforms = platforms.all;
+  meta = iosevka.meta // {
     maintainers = [ maintainers.cstrahan ];
   };
 
diff --git i/pkgs/data/fonts/iosevka/default.nix w/pkgs/data/fonts/iosevka/default.nix
index 9ef751688db..ef2df780e83 100644
--- i/pkgs/data/fonts/iosevka/default.nix
+++ w/pkgs/data/fonts/iosevka/default.nix
@@ -75,7 +75,7 @@ stdenv.mkDerivation rec {
 
   buildPhase = ''
     runHook preBuild
-    npm run build --no-update-notifier -- --jCmd=$NIX_BUILD_CORES ttf::$pname >/dev/null
+    npm run build --no-update-notifier -- --jCmd=$NIX_BUILD_CORES ttf::$pname
     runHook postBuild
   '';
 
```

1. Don't silence the output of `npm run build` - it's interesting IMO.
2. Don't repeat yourself (meta information).

**Comment by doronbehar** [comment]
> Should I drop the last commit (that updates node-packages.nix) even if it breaks `iosevka` on master in the meantime?

No, we can't break it directly on `master`. If there was an upcoming nodePackages `./generate.sh` update, it would have been appropriate to tell the updater to grab the commits from here and we'd close this PR.

**Comment by AluisioASG** [comment]
> Don't silence the output of `npm run build` - it's interesting IMO.

As far as I can tell there is no way to get a useful log without getting progress spam in the way, and any redirects just shut down all logging:

```
> iosevka@3.7.1 build /build/Iosevka-3.7.1
> verda -f verdafile.js "--jCmd=2" "--verbosity=6" "ttf::iosevka"
⢀⠀ ● Note Invalidating build journal due to self-tracking.
⡀⠀  | CPU:000% - RAM:000%⡀⠀ Complete 04 - Executing 2 - Pending 59 (06%) | CPU:000% - RAM:000%⠄⠀ Complete 004 - Executing 0 - Pending 110 (03%) | CPU:000% - RAM:000%⢂⠀ Complete 059 - Executing 2 - Pending 109 (34%) | CPU:000% - RAM:000%⡂⠀ Complete 062 - Executing 2 - Pending 161 (27%) | CPU:000% - RAM:000%⠅⠀ Complete 063 - Executing 0 - Pending 167 (27%) | CPU:000% - RAM:000%♦ node ./node_modules/patel/bin/patel-c --strict font-src/kits/boole-kit.ptl -o
⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%♦ node ./node_modules/patel/bin/patel-c --strict font-src/kits/spiro-kit.ptl -o
⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⡃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⠍⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⢋⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:053%⡋⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%⠍⠁ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%⢋⠁ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%♦ node ./node_modules/patel/bin/patel-c --strict font-src/meta/aesthetics.ptl
^C
```

Even if I reduce verbosity, it'll keep saving the progress bar updates for printing later.

```
> iosevka@3.7.1 build /build/Iosevka-3.7.1
> verda -f verdafile.js "--jCmd=2" "--verbosity=5" "ttf::iosevka"

⢀⠀ ● Note Invalidating build journal due to self-tracking.
```
When that build finished, it produced 7200 lines of the following:
Progress bar spam after build

**Comment by AluisioASG** [comment]
Sharing `meta` seems like it'll mess up with iosevka-bin's update script, as it locates the file to update with

https://github.com/NixOS/nixpkgs/blob/f28c4a120f5cacdad318782ccdd5b3599cea7e65/pkgs/data/fonts/iosevka/update.sh#L16

and what I get when doing `meta = iosevka.meta // ...` is

```
nix-repl> iosevka-bin.meta.position
"/home/aasg/nix/nixpkgs/iosevka/pkgs/data/fonts/iosevka/default.nix:98"
```

@deviant is there a problem if I replace that line with the following, or similar (I see you also write to variants.nix further down)?

```sh
file=$(dirname "${BASH_SOURCE[0]}")/bin.nix
```

**Comment by AluisioASG** [comment]
Meanwhile, I removed otfcc from the dependencies, added myself as maintainer, and updated the nodePackages update.

**Comment by deviant** [comment]
> Sharing `meta` seems like it'll mess up with iosevka-bin's update script, as it locates the file to update with
> 
> https://github.com/NixOS/nixpkgs/blob/f28c4a120f5cacdad318782ccdd5b3599cea7e65/pkgs/data/fonts/iosevka/update.sh#L16
> 
> @deviant is there a problem if I replace that line with the following, or similar (I see you also write to variants.nix further down)?
> 
> ```shell
> file=$(dirname "${BASH_SOURCE[0]}")/bin.nix
> ```

Not by me. I pretty much just copied that out of the common updater script. Perhaps it would be good to move iosevka-bin into a directory of its own, however? That way the file names can be more normal.

**Comment by AluisioASG** [comment]
> Perhaps it would be good to move iosevka-bin into a directory of its own, however? That way the file names can be more normal.

Possibly. The packages have little in common by now and we already do that with e.g. Firefox.

**Comment by doronbehar** [comment]
Sharing the meta should not be blocked by the design of the update script. You can put the version in a separate file and not use `nix-instantiate` or whatever it is.

@AluisioASG I'm sorry this got held up, but now we have a merge conflict...

If you could please change the update script to not rely on `meta.position` and use a safer way to increment the version, it'd be great.

**Comment by doronbehar** [comment]
As for the merge conflict, I don't know what to do. It's indeed frustrating. We need to coordinate runs of node's `./generate.sh`... (cc @cideM).

**Comment by doronbehar** [comment]
We could also (after the version increment issue will be solved) kindly ask @cideM to grab the commits from here (and perhaps along [this](https://github.com/NixOS/nixpkgs/commit/db7a346d485b890fa68424880fb88368d1615cdf) commit [please](https://discourse.nixos.org/t/call-for-nodepackages-maintainers/9176/27) :pray:) and then put them both in a separate PR and then run `./generate.sh`.

**Comment by AluisioASG** [comment]
I changed the update script, updating nodePackages now, should be done within the hour.

**Review by doronbehar — APPROVED** [review]
Well done.

File: pkgs/data/fonts/iosevka/update-bin.sh
Comment (deviant) [inline pkgs/data/fonts/iosevka/update-bin.sh:9]: This seems wrong; you're supposed to run the update script from the root directory of nixpkgs.

File: pkgs/data/fonts/iosevka/update-default.sh
Comment (deviant) [inline pkgs/data/fonts/iosevka/update-default.sh:6]: Same here.

**Comment by deviant** [comment]
You'll note that the [nerdfonts update script you mentioned](https://github.com/NixOS/nixpkgs/blob/a7f375d3354af159eab45896e012fb9e856b93cb/pkgs/data/fonts/nerdfonts/update.sh) does not actually cd into its directory.

File: pkgs/data/fonts/iosevka/update-bin.sh
Comment (doronbehar) [reply pkgs/data/fonts/iosevka/update-bin.sh:9]: The script changes the directory of itself before executing these commands.

Comment (deviant) [reply pkgs/data/fonts/iosevka/update-bin.sh:9]: It seems, at least, uncommon:

```
v@february ~/s/nixpkgs (local)> rg -t sh 'import \./' | wc -l
10
v@february ~/s/nixpkgs (local)> rg -t sh 'import \.\./' | wc -l
3
```

**Pull request merged**
